simplified-books-time-tracking/src/main/java/org/nypl/simplified/books/time/tracking/TimeTrackingService.kt

```diff
@@ -0,0 +1,399 @@
+    val libraryFile = File(timeTrackingDirectory, accountID.uuid.toString())
+
+    // create a directory for the library
+    if (!libraryFile.exists()) {
+      libraryFile.mkdirs()
+    }
+
+    val bookFile = File(libraryFile, bookId)
+
+    // create a directory for the book inside the library
+    if (!bookFile.exists()) {
+      bookFile.mkdirs()
+    }
+
+    timeEntriesFile = File(bookFile, FILE_NAME_TIME_ENTRIES)
+    timeEntriesToRetryFile = File(bookFile, FILE_NAME_TIME_ENTRIES_RETRY)
+
+    if (!timeEntriesFile.exists()) {
+
+      // create an entries file for this book with an initial time tracking info
+      timeEntriesFile.createNewFile()
+
+      TimeTrackingInfoFileUtils.saveTimeTrackingInfoOnFile(
+        timeTrackingInfo = TimeTrackingInfo(
+          accountId = accountID.uuid.toString(),
+          bookId = bookId,
+          libraryId = libraryId,
+          timeEntries = listOf(),
+          timeTrackingUri = timeTrackingUri
+        ),
+        file = timeEntriesFile
+      )
+    }
+
+    if (!timeEntriesToRetryFile.exists()) {
+
+      // create a file for possible entries that weren't successfully saved on the server
+      timeEntriesToRetryFile.createNewFile()
+
+      TimeTrackingInfoFileUtils.saveTimeTrackingInfoOnFile(
+        timeTrackingInfo = TimeTrackingInfo(
+          accountId = accountID.uuid.toString(),
+          bookId = bookId,
+          libraryId = libraryId,
+          timeEntries = listOf(),
+          timeTrackingUri = timeTrackingUri
+        ),
+        file = timeEntriesFile
+      )
+    }
+
+    isOnAudiobookScreen = true
+  }
+
+  override fun onPlayerEventReceived(playerEvent: PlayerEvent) {
+    when (playerEvent) {
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackProgressUpdate,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackStarted -> {
+        isPlaying = true
+        if (audiobookPlayingDisposable == null) {
+          createTimeTrackingEntry()
+          startEventDisposables()
+        }
+      }
+
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackBuffering,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackWaitingForAction,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventChapterWaiting,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackPaused,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventPlaybackStopped,
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventChapterCompleted -> {
+        isPlaying = false
+      }
+      is PlayerEvent.PlayerEventWithSpineElement.PlayerEventCreateBookmark,
+      is PlayerEvent.PlayerEventPlaybackRateChanged,
+      is PlayerEvent.PlayerEventError,
+      PlayerEvent.PlayerEventManifestUpdated -> {
+        // do nothing
+      }
+    }
+  }
+
+  override fun stopTracking() {
+    logger.debug("Stop tracking playing time")
+
+    audiobookPlayingDisposable?.dispose()
+    remoteStorageDisposable?.dispose()
+
```

Comment:
Might consider using the `CompositeDisposable` here. It's a `Disposable` in rxjava that you can add other disposables into. When you dispose of the composite, it disposes of all of the others, even if one of the disposables raises an exception.

Reply:
Yeah, I know. I didn't use the `CompositeDisposable` because I use one of the disposables more than once (to compare, set as null, reinitialize) so I didn't find the need of adding another object just to dispose them. However, since the `remoteStorageDisposable` variable is just used to be disposed, I can add the `CompositeDisposable` to handle it and drop this one.